File: solutions/c/isogram/1/isogram.c
#include "isogram.h"
#include <string.h>

bool is_isogram(const char phrase[]){

    if(!phrase)
        return false;
    size_t len = strlen(phrase);
    bool found_letter[26] = {0};
    bool is_isogram = true;

    if(0==len)
        return is_isogram;
    for(size_t i=0; i<len; i++){
        char c = phrase[i];
        int idx = 0;
        if(c>= 'a' && c<='z')
            idx = c-96;
        else if(c>= 'A' && c<='Z')
            idx = c-64;
        if (idx>0){   
            if(!found_letter[idx-1])
                found_letter[idx-1] = true; 
            else {
                 is_isogram = false;
                 break;
            }}
        }            
    return is_isogram;
}

File: solutions/free-pascal/isogram/2/Isogram.pas
unit Isogram;

{$mode ObjFPC}{$H+}

interface

function IsIsogram(const phrase : string) : boolean;

implementation

uses SysUtils;

function IsIsogram(const phrase : string) : boolean;
var 
  i, lt: integer;
  c: char;
  table : array[0..25] of boolean;
begin
  for i:=0 to 25 do
  begin
    table[i] := false;
  end;
  
  for c in phrase do
  begin
    lt := ord(c);
    if (lt < ord('A')) or ((lt>ord('Z')) and (lt<ord('a')))or (lt>ord('z')) then
      continue;
    if (lt>=ord('A')) and (lt<=ord('Z')) then
      lt:= lt - ord('A')
    else 
      lt:= lt - ord('a');  
    
    if not table[lt] then 
      table[lt] := true
    else
      exit(false);
  
  end;
  result := true;
end;

end.
